streams/src/test/java/org/apache/kafka/streams/processor/internals/TaskManagerTest.java

```diff
@@ -1804,25 +1805,52 @@ public void shouldMaybeCommitAllActiveTasksThatNeedCommit() {
     @Test
     public void shouldProcessActiveTasks() {
         final StateMachineTask task00 = new StateMachineTask(taskId00, taskId00Partitions, true);
+        final StateMachineTask task01 = new StateMachineTask(taskId01, taskId01Partitions, true);
+
+        final Map<TaskId, Set<TopicPartition>> assignment = new HashMap<>();
+        assignment.put(taskId00, taskId00Partitions);
+        assignment.put(taskId01, taskId01Partitions);
 
         expectRestoreToBeCompleted(consumer, changeLogReader);
-        expect(activeTaskCreator.createTasks(anyObject(), eq(taskId00Assignment)))
-            .andReturn(singletonList(task00)).anyTimes();
+        expect(activeTaskCreator.createTasks(anyObject(), eq(assignment)))
+            .andReturn(Arrays.asList(task00, task01)).anyTimes();
 
         replay(activeTaskCreator, consumer, changeLogReader);
 
-        taskManager.handleAssignment(taskId00Assignment, emptyMap());
+        taskManager.handleAssignment(assignment, emptyMap());
         assertThat(taskManager.tryToCompleteRestoration(), is(true));
 
         assertThat(task00.state(), is(Task.State.RUNNING));
+        assertThat(task01.state(), is(Task.State.RUNNING));
 
-        final TopicPartition partition = taskId00Partitions.iterator().next();
         task00.addRecords(
-            partition,
-            singletonList(new ConsumerRecord<>(partition.topic(), partition.partition(), 0L, null, null))
+            t1p0,
+            Arrays.asList(
+                getConsumerRecord(t1p0, 0L),
+                getConsumerRecord(t1p0, 1L),
+                getConsumerRecord(t1p0, 2L),
+                getConsumerRecord(t1p0, 3L),
+                getConsumerRecord(t1p0, 4L),
+                getConsumerRecord(t1p0, 5L)
+            )
         );
+        task01.addRecords(
+            t1p1,
+            Arrays.asList(
+                getConsumerRecord(t1p1, 0L),
+                getConsumerRecord(t1p1, 1L),
+                getConsumerRecord(t1p1, 2L),
+                getConsumerRecord(t1p1, 3L),
+                getConsumerRecord(t1p1, 4L)
+            )
+        );
+
+        // check that we should be processing at most max num records
+        assertThat(taskManager.process(3, 0L), is(6));
 
-        assertThat(taskManager.process(0L), is(1));
+        // check that if there's no records proccssible, we would stop early
```

Comment:
Not sure if there's a better way to put it, but I was a little surprised to see "we would stop early". I guess it didn't even occur to me that the task manager would block or something if there's no records to process.

Reply:
Oh maybe I did not put it clearly: let's say for a single task, if we call process(N) but there are only M < N records in the buffer, then of course we would process only M records.

If there are multiple tasks, say, 2, if we call process(N) and task1 has N1 > N and task2 has N2 < N, then we would process a total of N + N2 in a single call, and in that case we would not "stop" but continues to process; we would only break the loop when ALL tasks have no records to process. This test is just to verify that part.

Reply:
Ah, ok... To be clear, I was only confused about the comment, not the behavior. I think the comment was "obvious" to me only because this is exactly what this PR is about. But coming back in a year, I'm sure I'll appreciate the comment.